File: haskell/src/Solutions/Euler306.hs
module Solutions.Euler306 where

-- http://oeis.org/A215721
-- Found using simulation
-- The most important line is:
-- For n > 14, a(n) = a(n - 5) + 34

initSequence, period, totalSequence :: [Int]
initSequence = [1, 5, 9, 15, 21, 25, 29, 35]
period = [39, 43, 55, 59, 63]
totalSequence = initSequence ++ concatMap (\s -> map (+s) period) [0, 34..]

funF :: Int -> Int
funF n = n - length (takeWhile (<=n) totalSequence)

euler306 :: IO String
euler306 = return $ show $ funF $ 10^6

File: rust/src/solutions/euler306.rs
// http://oeis.org/A215721
// Found using simulation
// The most important line is:
// For n > 14, a(n) = a(n - 5) + 34

fn f(n: usize) -> usize {
    let init_seq = vec![1usize, 5, 9, 15, 21, 25, 29, 35];
    let period = vec![39usize, 43, 55, 59, 63];
    let and_then = (0..)
        .step_by(34)
        .flat_map(|s| period.iter().map(move |&t| s + t));
    let total_seq = init_seq.into_iter().chain(and_then);
    n - total_seq.take_while(|&x| x <= n).count()
}

pub fn euler306() -> String {
    f(1_000_000).to_string()
}
